keystone/api/projects.py

```diff
@@ -47,6 +48,49 @@ def _build_project_target_enforcement():
     return target
 
 
+def is_provider_tag(tag: str) -> bool:
+    """Check if the provided tag is a provider tag."""
+    return tag.startswith(tuple(CONF.provider_tag_prefix))
+
+
+def get_provider_tags(tags: Iterable[str]) -> set[str]:
+    """Return the provider tags from the given tags."""
+    return set(t for t in tags if is_provider_tag(t))
+
+
+def validate_tags(project_id: str, tags: Iterable[str], policy: str
+                  ) -> list[str]:
+    """Return the tags the user is allowed to change.
+
+    We check whether the list of tags contains provider tags. If it does, we
+    check if the user is allowed to change provider tags. If they are not
+    allowed, we restore the original provider tags in the list.
+    """
+    original_ptags = get_provider_tags(
+        PROVIDERS.resource_api.list_project_tags(project_id)
+    )
+    ptags = get_provider_tags(tags)
+    removed_ptags = original_ptags - ptags
+    new_ptags = ptags - original_ptags
+    if removed_ptags or new_ptags:
+        # NOTE(jkulik): Adding/removing provider tags needs special
+        # privileges but the user might just send us an update of their own
+        # tags instead of including the provider tags, so we try to handle
+        # this gracefully here by updating the non-provider tags
+        # nonetheless.
+        try:
+            ENFORCER.enforce_call(
+                action='identity:%s:provider_tags' % policy,
+                build_target=_build_project_target_enforcement
+            )
+        except exception.Forbidden:
+            # User is not allowed to change provider tags so we restore the
+            # original provider tags to allow them to update their tags.
+            tags = set(tags) + removed_ptags - new_ptags
```

Comment:
`+` needs to be changed to `|` i think. After i did it in a local environment it worked:
```
(set(tags) | removed_ptags) - new_ptags
```